### PR TITLE
Feat: hidden-fee-annotate rule for drip-pricing fees (#119)

### LINKED ISSUE #119: Rule proposal: hidden-fee-annotate — annotate drip-pricing fees added during checkout
### Category
New defense rule

### What problem does this solve?
Browser-use agents asked to "buy X for \$Y" on hotel, ticket, travel, and e-commerce sites encounter the *drip pricing* pattern: a low base price up front, with mandatory fees ("service fee", "resort fee", "convenience fee", "processing fee", "delivery surcharge") only added at checkout. The FTC's [Trade Regulation Rule on Unfair or Deceptive Fees](https://www.ftc.gov/news-events/news/press-releases/2024/12/federal-trade-commission-announces-bipartisan-rule-banning-junk-ticket-hotel-fees) (effective 2025-05-12) bans the pattern for live-event tickets and short-term lodging, but enforcement is case-by-case elsewhere and the pattern remains common across airlines, food delivery, and event-adjacent commerce.

The asymmetry an agent faces: a sighted user scans the running total before clicking Pay; an agent acting on a single task instruction may not re-check pricing assumptions at the final step. The agent isn't pressured by manufactured urgency (`countdown-timer-redact` territory) — it's misled by the order in which costs become visible.

### Proposed solution
On checkout-shaped URLs (reuse the gate `cart-addon-annotate` and `checkout-checkbox-sanitize` already use: `/cart`, `/checkout`, `/basket`, `/bag`, `/payment`, `/order`), walk line-item rows and annotate ones whose visible label matches a curated hidden-fee phrase set ("service fee", "convenience fee", "processing fee", "resort fee", "destination fee", "facility fee", "handling fee", "venue fee", "delivery surcharge"). Annotation reuses the visible inline-chip pattern from `cart-addon-annotate` — the line item is preserved so the agent decides whether to abandon, downgrade, or proceed.

**MVP:** label-only matching. No price comparison, no anchor selection.

The hardest part is choosing a comparison anchor for any future "delta vs. advertised" check — possible anchors are referrer-URL price params, the cart page's initial total snapshot, or an explicit "starting at" / "from" element earlier in the flow. Defer past v1.

### Alternatives considered
- **Strip the fee row outright.** Wrong choice — many fees are real (taxes, shipping). The defense is making the agent *see* the fee, not hiding it. One of the few cases where annotate is the right action even under a remove-over-annotate default; the rule must not destroy information the operator is legally required to surface.
- **Compute a delta vs. some baseline total.** Anchor selection is the open problem and varies by site; ship label-flag first and iterate.
- **Defer to `irrelevant-sections-redact`.** Wrong threat model — fees are not irrelevant chrome, they're load-bearing pricing.

### Controlling false positives
- **Shape-check the row.** Matching label must sit in an order-summary container (`<table>` / `[role="region"]` / `<aside>` with cart-shaped class names / schema.org `Order` microdata) AND have an adjacent currency amount within N siblings. Loose paragraph mentions of "service fee" in marketing copy do not match.
- **Whole-string regex on the label element**, same precision bar as `disguised-ad-flag`. Substring matches like "Customer Service Fee Schedule" inside a sentence get rejected.
- **Explicit exclude list for legally-required line items** — "tax", "VAT", "GST", "sales tax", "shipping", "delivery" without "surcharge". These already have their own legal disclosure regime and are not the hidden-fee pattern.
- **Single-item cart exclusion.** When the matched fee row is the *only* item in the cart, it's the product, not an add-on (e.g., paying a service fee directly to a utility).
- **Per-host kill-switch.** Sites where the fee genuinely is the product (DMV portals, court e-filing) can be denylisted.
- **Phrase set curation policy.** Additions go through PR review; avoid expanding to single overly-generic words like "fee" alone. Same governance posture as `cart-addon-annotate`'s pattern set.
- **Telemetry on rollout.** Default-off until per-rule hit/skip counts on the live ruleset show the FP rate is acceptable; promote to default-on after a real-world signal window similar to `schema-trust-sanitize`'s posture.

### Prior art / references
- FTC Trade Regulation Rule on Unfair or Deceptive Fees ([final rule PDF](https://www.ftc.gov/system/files/ftc_gov/pdf/r207011_udf_rule_2024_final_0.pdf)).
- Mathur et al. (CSCW 2019), *Dark Patterns at Scale.* — covers *Sneaking* and *Hidden Costs*. Already cited by other rules.
- Brignull, [deceptive.design](https://www.deceptive.design/) — *Hidden Costs* and *Sneak into Basket*.
- No known direct OSS implementation. Currency parsing handled by built-in `Intl.NumberFormat` or [`currency.js`](https://github.com/scurker/currency.js) (MIT).

Tagged Impact M–H / Complexity H.

### PR DESCRIPTION
**Pull request by twschiller**
## Summary

- New rule `hidden-fee-annotate` (default **off**, Experimental) — annotates mandatory fees that surface only at checkout (resort / service / convenience / processing / destination / facility / handling / venue fee, delivery surcharge) so agents can verify the total before paying. Closes #119.
- Layered false-positive control: whole-string regex on a small leaf-ish label, required order-summary ancestor (`<table>` / `[role="region"]` / cart-shaped class/id / schema.org `Order` microdata), required adjacent currency amount, single-item-cart skip, legally-required-line-item exclude list (tax, VAT, GST, shipping, tip, gratuity), and a (launch-empty) per-host denylist.
- Row is annotated, not removed — many of these fees are legally required to be surfaced (FTC Unfair-or-Deceptive-Fees rule, effective 2025-05-12).
- Test coverage: 70 example-based tests + 8 `fast-check` property tests (regex precision, phrase/exclude disjointness, currency regex, single-item-cart invariant, idempotency). Full suite: 1444 tests pass.

## Rollout

Ships default-off; promote to default-on after per-rule activity counts (#174) show the FP rate is acceptable, mirroring `schema-trust-sanitize`'s posture.

## Test plan

- [x] `bun run preflight` (build-site-data, build-injection-patterns, build-rule-defaults, biome, eslint, tsc, knip, jest --coverage) — clean
- [x] `pre-commit run --files docs/src/content/docs/rules.md` — clean (mdformat-reflowed on first run)
- [ ] Manual smoke test: enable the rule, load a hotel checkout with a known resort fee, confirm the chip appears on the resort-fee row only and the popup's per-rule count increments
- [ ] Manual smoke test: confirm the rule does not annotate marketing-copy mentions of "service fee" on a non-checkout URL or on a `/checkout` page outside the order-summary container

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: extension/src/rules/hidden-fee-annotate.ts [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+    return direct;
+  }
+  const labelledby = element.getAttribute("aria-labelledby");
+  if (labelledby === null || labelledby.length === 0) {
+    return "";
+  }
+  // aria-labelledby can list multiple IDs; concatenate their visible text.
+  const document_ = element.ownerDocument;
+  const parts: string[] = [];
+  for (const id of labelledby.split(/\s+/)) {
```
[the remark this answers is not in the document]
Comment (twschiller) [reply extension/src/rules/hidden-fee-annotate.ts:235]: Good catch — fixed in 5af258b. Added the empty-id guard alongside the `split(/\s+/)`, matching `readAccessibleName` in `trust-badge-annotate.ts` as suggested. Also added a regression test for the leading/trailing-whitespace case in `hidden-fee-annotate.test.ts` (it needs a small `CSS.escape` polyfill because jsdom in this project doesn't expose the `CSS` global — same latent test-only quirk applies to trust-badge but isn't exercised there).

— Claude Code, on behalf of @twschiller
